Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp

```diff
@@ -107,6 +106,40 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
 }
 #endif
 
+ScrollingStateTree& AsyncScrollingCoordinator::ensureScrollingStateTreeForRootFrameID(FrameIdentifier identifier)
+{
+    auto it = m_scrollingStateTrees.find(identifier);
+    if (it != m_scrollingStateTrees.end())
+        return it->value.get();
+
+    ASSERT_WITH_MESSAGE(!page() || page()->settings().siteIsolationEnabled() || m_scrollingStateTrees.isEmpty(), "With site isolation off, we should only ever have one scoling state tree. Get the frame identifier from LocalFrame::rootFrame.");
+    auto addResult = m_scrollingStateTrees.set(identifier, makeUniqueRef<ScrollingStateTree>(this));
+    return addResult.iterator->value.get();
+}
+
+void AsyncScrollingCoordinator::removeRootFrame(FrameIdentifier identifier)
+{
+    m_scrollingStateTrees.remove(identifier);
+}
+
+RefPtr<ScrollingStateNode> AsyncScrollingCoordinator::stateNodeForID(ScrollingNodeID nodeID) const
```

Comment:
`stateNodeForNodeID()` to disambiguate from other kinds of IDs.

I'm not really a fan of having to do lookups in all the scrolling trees, since this gets called a lot. In many call sites we might already have the root frame identifier?

Reply:
I think it's preferable to do this versus having to go through frameViewForScrollingNode. For the call sites with the frame view we can maybe get away with getting the root frame id that way.

Reply:
After https://github.com/WebKit/WebKit/pull/27823 added roofFrameID to ScrollableArea, so for cases where we have the ScrollableArea we can avoid looking through the scrolling trees.